feat: typescript migration

SlayerOrnstein opened the pull request:
### What did you fix? <!-- provide a description or issue closes statement -->
Migrates code base to typescript. Still messing around with it since linter is screaming at using `any`

---

### Reproduction steps
<!--
1. I did the thing
1. Then I clicked the button
1. Then I deleted the word
1. Then I found the error!
-->

---

### Evidence/screenshot/link to line

<!-- You can see my fix [on this line](#line-number-1235234) for where the new data exists -->

### Considerations
- Does this contain a new dependency? **Yes**
- Does this introduce opinionated data formatting or manual data entry? **Yes**
- Does this pr include updated data files in a separate commit that can be reverted for a clean code-only pr? **No**
- Have I run the linter? **Yes**
- Is is a bug fix, feature request, or enhancement? **Feature**

TobiTenno commented on exports.ts:
```diff
@@ -0,0 +1,148 @@
+import safeImport from './safeImport.ts';
+import { Arcane, ArchonShard, Conclave, Events, SolNode, SortieData, SteelPath, SynthesisTarget } from './types.ts';
+
+const locales = ['de', 'es', 'fr', 'it', 'ko', 'pl', 'pt', 'ru', 'zh', 'cs', 'sr', 'uk'];
```
no locale type?

TobiTenno commented on test/utilities/timedate.spec.ts:
```diff
@@ -27,8 +22,6 @@ describe('timeDateUtils', () => {
   describe('parseDate()', () => {
     it('should parse even if date provided is undefined', () => {
       (() => parseDate()).should.not.throw();
-      (() => parseDate(0)).should.not.throw();
-      (() => parseDate(2340985790347890)).should.not.throw();
```
why're these removed? we discussed that js consumers can still use invalid types

TobiTenno commented on biome.json:
kindly nuke the tabs, please....

also, not sure why this is so gorram long

TobiTenno commented on package.json:
```diff
@@ -39,47 +39,59 @@
-    ".": "./exports.js",
-    "./utilities": "./tools/utilities.js"
+    ".": "./exports.ts",
+    "./utilities": "./tools/utilities.ts",
+    "./types": "./types.ts"
   },
-  "main": "exports.js",
+  "main": "exports.ts",
```
these should all be paths to dist js files

TobiTenno commented on package.json:
```diff
@@ -39,47 +39,59 @@
+    }
+  },
+  "dependencies": {
+    "@types/yargs": "^17.0.33"
```
dev dependency, don't need the types for anything besides tools, which isn't exported, right?

TobiTenno commented on test/exports.spec.ts:
```diff
@@ -1,22 +1,24 @@
+// biome-ignore-all lint/suspicious/noConsole: calls to the console are only used for test
```
with all those lines in the biome config, we can't ignore a directory?

TobiTenno commented on test/utilities/integrity.spec.js:
it says it's empty now, probably delete?

TobiTenno commented on package.json:
```diff
@@ -39,11 +39,11 @@
+    "./types": "./dist/types.ts"
   },
-  "main": "exports.ts",
+  "main": "./dist/exports.ts",
```
again, these should be dist js files...

TobiTenno commented on biome.json:
```diff
@@ -0,0 +1,773 @@
+                            "frameElement": "TODO: Add a custom message here.",
+                            "scrollY": "TODO: Add a custom message here.",
+                            "self": "TODO: Add a custom message here.",
+                            "onblur": "TODO: Add a custom message here.",
+                            "find": "TODO: Add a custom message here.",
+                            "parent": "TODO: Add a custom message here.",
+                            "top": "TODO: Add a custom message here.",
+                            "moveBy": "TODO: Add a custom message here.",
+                            "menubar": "TODO: Add a custom message here.",
+                            "length": "TODO: Add a custom message here.",
+                            "onerror": "TODO: Add a custom message here.",
+                            "onresize": "TODO: Add a custom message here.",
+                            "removeEventListener": "TODO: Add a custom message here.",
+                            "onload": "TODO: Add a custom message here.",
+                            "scrollTo": "TODO: Add a custom message here.",
+                            "moveTo": "TODO: Add a custom message here.",
+                            "scrollX": "TODO: Add a custom message here.",
+                            "name": "TODO: Add a custom message here.",
+                            "toolbar": "TODO: Add a custom message here.",
+                            "innerWidth": "TODO: Add a custom message here.",
+                            "location": "TODO: Add a custom message here.",
+                            "locationbar": "TODO: Add a custom message here.",
+                            "scrollBy": "TODO: Add a custom message here.",
+                            "resizeTo": "TODO: Add a custom message here.",
+                            "stop": "TODO: Add a custom message here.",
+                            "scrollbars": "TODO: Add a custom message here.",
+                            "blur": "TODO: Add a custom message here.",
+                            "screenTop": "TODO: Add a custom message here.",
+                            "confirm": "TODO: Add a custom message here.",
+                            "screen": "TODO: Add a custom message here.",
+                            "screenLeft": "TODO: Add a custom message here.",
+                            "event": "TODO: Add a custom message here.",
+                            "onunload": "TODO: Add a custom message here.",
+                            "pageXOffset": "TODO: Add a custom message here.",
+                            "resizeBy": "TODO: Add a custom message here.",
+                            "statusbar": "TODO: Add a custom message here.",
+                            "isNaN": "Use Number.isNaN instead https://github.com/airbnb/javascript#standard-library--isnan",
+                            "close": "TODO: Add a custom message here.",
+                            "open": "TODO: Add a custom message here.",
+                            "print": "TODO: Add a custom message here."
```
maybe biome isn't as ready from migrations as i thought, cause this is... excessive.

TobiTenno commented on biome.json:
```diff
@@ -0,0 +1,773 @@
+        },
+        {
+            "includes": [
+                "*.ts",
+                "*.tsx",
+                "*.mts",
+                "*.cts"
+            ],
+            "linter": {
+                "rules": {
+                    "complexity": {
+                        "noArguments": "error"
+                    },
+                    "correctness": {
+                        "noConstAssign": "off",
+                        "noGlobalObjectCalls": "off",
+                        "noInvalidBuiltinInstantiation": "off",
+                        "noInvalidConstructorSuper": "off",
+                        "noSetterReturn": "off",
+                        "noUndeclaredVariables": "off",
+                        "noUnreachable": "off",
+                        "noUnreachableSuper": "off"
+                    },
+                    "style": {
+                        "useConst": "error"
+                    },
+                    "suspicious": {
+                        "noDuplicateClassMembers": "off",
+                        "noDuplicateObjectKeys": "off",
+                        "noDuplicateParameters": "off",
+                        "noFunctionAssign": "off",
+                        "noImportAssign": "off",
+                        "noRedeclare": "off",
+                        "noUnsafeNegation": "off",
+                        "noVar": "error",
+                        "useGetterReturn": "off"
+                    }
+                }
+            }
+        },
```
several of these seem like duplicates, please try deduping the array

SlayerOrnstein commented:
> overall fine, i'm just getting skeptical of biome with so much duping

Yeah, I just removed everything and use the recommend for all with some tweaks to work with this project specifically

Looks like the migration tool biome has also followed and added all the eslint extensions

TobiTenno commented on biome.json:
```diff
@@ -31,737 +31,25 @@
-                "useDefaultSwitchClauseLast": "error",
-                "useGetterReturn": "error",
-                "useGuardForIn": "error"
+                "noExplicitAny": "off"
```
i'd actually like this one on, personally, as a warning, at least

TobiTenno commented on biome.json:
```diff
@@ -10,7 +10,7 @@
+        "ignoreUnknown": true
     },
     "formatter": {
         "enabled": true,
```
does this formatter use prettier? if not, we might not want this?

TobiTenno replied:
ahhh, i'd rather have prettier take precedence, so let's do this off

TobiTenno commented on .vscode/settings.json:
i'd rather not have IDE settings in the project, is there any other way? mostly cause i hate vscode

TobiTenno commented on .github/workflows/release.yaml:
```diff
@@ -50,6 +50,7 @@ jobs:
       - uses: actions/setup-node@v4
         with:
           node-version-file: '.nvmrc'
+      - run: npm run build
```
build before install? that seems wrong. you probably want to just add `build` to `prepublishOnly` instead.

TobiTenno reviewed (changes requested):
once you have all these changes in, could you do an integration build w/ parser to make sure it's passive?

TobiTenno commented on tools/integrity.ts:
```diff
@@ -3,7 +3,7 @@
  * @param args arguments to ensure
  */
-export const insist = (thing: Record<string, any>, args: string[]) => {
+export const insist = (thing: Record<string, unknown>, args: string[]) => {
```
:+1:

SlayerOrnstein commented:
@TobiTenno 
test log for:

[warframe-status](https://github.com/user-attachments/files/21026959/logs.txt)
[worldstate-parser](https://github.com/user-attachments/files/21026987/logs.txt)
[warframe-items](https://github.com/user-attachments/files/21027002/logs.txt)
